### PR TITLE
Add product variants events

### PR DESCRIPTION
**Pull request by piotrgrundas**
I want to merge this change because I need events for product variant:

* [x] variant create
* [x] product variant update
* [x] product variant delete

[Dashboard PR](https://github.com/mirumee/saleor-dashboard/pull/1006)

# Impact

* [ ] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [x] Privileged queries and mutations are guarded by proper permission checks
* [x] Database queries are optimized and the number of queries is constant
* [x] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [x] Changes are mentioned in the changelog

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ All notable, unreleased changes to this project will be documented in this file.
 - Deprecate API fields `Order.discount`, `Order.discountName`, `Order.translatedDiscountName` - #6874 by @korycins
 - Fix argument validation in page resolver - #6960 by @fowczarek
 - Drop `data` field from checkout line model - #6961 by @fowczarek
+- Add PRODUCT_VARIANT_CREATED/UPDATED/DELETED webhooks, fix attributes field for PRODUCT_CREATED/UPDATED webhooks - #6963 by @piotrgrundas
```
Comment (korycins) [inline CHANGELOG.md:17]: Maybe `PRODUCT_VARIANT_CREATED`, `UPDATED`, `DELETED`

File: saleor/graphql/product/bulk_mutations/products.py [hunk 2]
```diff
@@ -500,6 +501,10 @@ def perform_mutation(cls, _root, info, ids, **data):
             .values_list("pk", flat=True)
         )
 
+        product_variants = models.ProductVariant.objects.filter(id__in=pks)
+        for product_variant in product_variants:
+            info.context.plugins.product_variant_deleted(product_variant)
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:506]: We should trigger webhook in `transaction.on_commit`.

File: saleor/graphql/product/mutations/products.py [hunk 3]
```diff
@@ -891,7 +892,12 @@ def save(cls, info, instance, cleaned_input):
         if attributes:
             AttributeAssignmentMixin.save(instance, attributes)
             generate_and_set_variant_name(instance, cleaned_input.get("sku"))
+
         info.context.plugins.product_updated(instance.product)
+        if new_variant:
+            info.context.plugins.product_variant_created(instance)
+        else:
+            info.context.plugins.product_variant_updated(instance)
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:900]: We should trigger webhook in `transaction.on_commit`.

File: saleor/graphql/product/mutations/products.py [hunk 4]
```diff
@@ -979,20 +985,20 @@ def success_response(cls, instance):
...
         # get draft order lines for variant
         line_pks = list(
             order_models.OrderLine.objects.filter(
-                variant__pk=variant_pk, order__status=OrderStatus.DRAFT
+                variant__pk=instance.pk, order__status=OrderStatus.DRAFT
             ).values_list("pk", flat=True)
         )
 
         response = super().perform_mutation(_root, info, **data)
+        info.context.plugins.product_variant_deleted(instance)
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:998]: We should trigger webhook in `transaction.on_commit`.

File: saleor/graphql/product/bulk_mutations/products.py [hunk 5]
```diff
@@ -431,6 +431,7 @@ def save_variants(cls, info, instances, product, cleaned_inputs):
             cls.save(info, instance, cleaned_input)
             cls.create_variant_stocks(instance, cleaned_input)
             cls.create_variant_channel_listings(instance, cleaned_input)
+            info.context.plugins.product_variant_created(instance)
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:434]: We should trigger webhook in `transaction.on_commit`.

File: saleor/webhook/tests/test_webhook_payloads.py [hunk 6]
```diff
@@ -82,3 +88,18 @@ def test_order_lines_have_all_required_fields(order, order_line_with_one_allocat
         ),
         "tax_rate": str(line.tax_rate.quantize(Decimal("0.0001"))),
     }
+
+
+def test_generate_product_variant_payload(product_with_variant_with_two_attributes):
```
Comment (fowczarek) [inline saleor/webhook/tests/test_webhook_payloads.py:93]: We should add test `test_generate_product_variant_payload_for_deleted_variant`

File: saleor/webhook/tests/test_webhook_payloads.py [hunk 7]
```diff
@@ -82,3 +88,18 @@ def test_order_lines_have_all_required_fields(order, order_line_with_one_allocat
...
+    variant_id = graphene.Node.to_global_id("ProductVariant", variant.id)
+    additional_fields = ["channel_listings"]
+    extra_dict_data = ["attributes", "product_id"]
+    payload_fields = chain(
+        *[PRODUCT_VARIANT_FIELDS, extra_dict_data, additional_fields]
+    )
+
+    assert variant_id == payload["id"]
+    for field in payload_fields:
+        assert payload.get(field) is not None
```
Comment (fowczarek) [inline saleor/webhook/tests/test_webhook_payloads.py:105]: I think we should check the values of these fields.

File: saleor/webhook/serializers.py [hunk 8]
```diff
@@ -28,3 +34,38 @@ def serialize_checkout_lines(checkout: "Checkout") -> List[dict]:
...
+
+
+def serialize_product_attributes(
+    product: Union["Product", "ProductVariant"]
+) -> List[dict]:
+    queryset = (
+        (
+            AssignedProductAttribute.objects.filter(product_id=product.id)
+            if isinstance(product, Product)
+            else AssignedVariantAttribute.objects.filter(variant_id=product.id)
```
Comment (IKarbowiak) [inline saleor/webhook/serializers.py:46]: Shouldn't be:
```suggestion
            else AssignedVariantAttribute.objects.filter(variant__product_id=product.id)
```

**Review by fowczarek — CHANGES_REQUESTED** [review]
We have a few more mutations to update ProductVariant:
- ProductVariantStocksCreate
- ProductVariantStocksDelete
- ProductVariantStocksUpdate
- ProductVariantTranslate
- ProductVariantChannelListingUpdate
- ProductVariantReorderAttributeValues
- VariantImageAssign
- VariantImageUnassign

I'm not sure that we should trigger webhook in all the above mutations but for sure we should trigger webhook in `ProductVariantChannelListingUpdate`

File: saleor/webhook/serializers.py (hunk 8 above)
Comment (piotrgrundas) [reply saleor/webhook/serializers.py:46]: No :) It's good. It's a Product or productVariant

File: saleor/webhook/serializers.py [hunk 9]
```diff
@@ -28,3 +34,38 @@ def serialize_checkout_lines(checkout: "Checkout") -> List[dict]:
             }
         )
     return data
+
+
+def serialize_product_attributes(
+    product: Union["Product", "ProductVariant"]
```
Comment (fowczarek) [inline saleor/webhook/serializers.py:40]: Maybe?
```suggestion
def serialize_product_or_variant_attributes(
    product_or_variant: Union["Product", "ProductVariant"]
```

File: saleor/webhook/serializers.py (hunk 8 above)
Comment (IKarbowiak) [reply saleor/webhook/serializers.py:46]: True, sorry :)

**Comment by piotrgrundas** [comment]
@fowczarek I would skip `ProductVariantStocks` since I don't pass stocks to webhook payload. Moreover we'll need sth with data how stock changed for notifications so it will be probably a separate webhook.
`ProductVariantReorderAttributeValues` i don't order it and I don't think this is important for external integration.
Rest is fine absolutely 🙆‍♂️

File: saleor/webhook/tests/test_webhook_payloads.py (hunk 6 above)
Comment (piotrgrundas) [reply saleor/webhook/tests/test_webhook_payloads.py:93]: What's the point here? It's exactly the same ATM.

File: CHANGELOG.md [hunk 10]
```diff
@@ -14,6 +15,7 @@ All notable, unreleased changes to this project will be documented in this file.
 - Deprecate API fields `Order.discount`, `Order.discountName`, `Order.translatedDiscountName` - #6874 by @korycins
 - Fix argument validation in page resolver - #6960 by @fowczarek
 - Drop `data` field from checkout line model - #6961 by @fowczarek
+- Add `PRODUCT_VARIANT_CREATED`, `UPDATED`, `DELETED` webhooks, fix attributes field for `PRODUCT_CREATED`, `UPDATED` webhooks - #6963 by @piotrgrundas
```
Comment (korycins) [inline CHANGELOG.md:18]: My bad that I didn't explicitly  provide you a proper sample. it should be `PRODUCT_VARIANT_CREATED`, `PRODUCT_VARIANT_UPDATED`, `PRODUCT_VARIANT_DELETED`

File: CHANGELOG.md [hunk 11]
```diff
@@ -43,6 +44,7 @@ All notable, unreleased changes to this project will be documented in this file.
     - `get_checkout_line_tax_rate`
     - `preprocess_order_creation`
   - additionally, `preprocess_order_creation` was extend with `lines_info` parameter
+- remove `product_updated` webhook in `ProductVariantCreate`, replace `product_updated` with `product_variant_updated` webhook in `ProductVariantChannelListingUpdate` #6963 by @piotrgrundas
```
Comment (korycins) [inline CHANGELOG.md:47]: Maybe something like this:
```
Remove triggering a webhook event `PRODUCT_UPDATED`  when calling `ProductVariantCreate` mutation.  Use `PRODUCT_VARIANT_CREATED` instead. 
Remove triggering a webhook event `PRODUCT_UPDATED` when calling  `ProductVariantChannelListingUpdate` mutation. Use `PRODUCT_VARIANT_UPDATED` instead. 
```
@maarcingebala  What do you think about this? You're better than me in formatting msg like this one

File: saleor/graphql/product/mutations/products.py [hunk 12]
```diff
@@ -891,7 +892,12 @@ def save(cls, info, instance, cleaned_input):
         if attributes:
             AttributeAssignmentMixin.save(instance, attributes)
             generate_and_set_variant_name(instance, cleaned_input.get("sku"))
-        info.context.plugins.product_updated(instance.product)
+
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:895]: It is only a detail but maybe soemthing more readable:
```
event_to_call = (
    info.context.plugins.product_variant_created if new_variant else 
    info.context.plugins.product_variant_updated
)
transaction.on_commit(lambda: event_to_call(instance))
```
@fowczarek  What do you think?

File: saleor/graphql/product/mutations/products.py [hunk 13]
```diff
@@ -979,20 +985,31 @@ def success_response(cls, instance):
...
         )
 
+        # Get cached variant with related fields to fully populate webhook payload.
+        variant = (
+            models.ProductVariant.objects.prefetch_related(
+                "channel_listings",
+                "attributes__values",
+            )
+        ).get(id=instance.id)
+        with transaction.atomic():
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:1004]: Hhm... It is strange. You're applying the transaction only to call transaction.on_commit 😄 
Also to be consistent with `product_deleted` event. You should call the event that variant was removed after we really remove it from DB.

File: saleor/graphql/product/mutations/products.py [hunk 14]
```diff
@@ -1484,6 +1501,12 @@ def perform_mutation(cls, _root, info, image_id, variant_id):
                     }
                 )
         variant = ChannelContext(node=variant, channel_slug=None)
+
+        with transaction.atomic():
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:1505]: See my comment above comment according to transaction.
Why do we call call variant_delete for mutation which assigns images?

File: saleor/graphql/product/mutations/products.py [hunk 15]
```diff
@@ -1530,4 +1553,10 @@ def perform_mutation(cls, _root, info, image_id, variant_id):
             variant_image.delete()
 
         variant = ChannelContext(node=variant, channel_slug=None)
+
+        with transaction.atomic():
+            transaction.on_commit(
+                lambda: info.context.plugins.product_variant_deleted(variant)
+            )
```
Comment (IKarbowiak) [inline saleor/graphql/product/mutations/products.py:1560]: The same situation as above, IMO `product_variant_deleted` shouldn't be called for `VariantImageUnassign` mutation. Rather, `product_variant_updated`.

File: saleor/graphql/product/mutations/products.py (hunk 12 above)
Comment (fowczarek) [reply saleor/graphql/product/mutations/products.py:895]: In my opinion, the solution proposed by @korycins is more readable.

File: saleor/graphql/product/mutations/products.py (hunk 15 above)
Comment (piotrgrundas) [reply saleor/graphql/product/mutations/products.py:1560]: Typo... 🙈

File: saleor/graphql/product/mutations/products.py (hunk 14 above)
Comment (piotrgrundas) [reply saleor/graphql/product/mutations/products.py:1505]: Typo

File: saleor/graphql/product/bulk_mutations/products.py [hunk 16]
```diff
@@ -431,6 +431,9 @@ def save_variants(cls, info, instances, product, cleaned_inputs):
             cls.save(info, instance, cleaned_input)
             cls.create_variant_stocks(instance, cleaned_input)
             cls.create_variant_channel_listings(instance, cleaned_input)
+            transaction.on_commit(
```
Comment (korycins) [inline saleor/graphql/product/bulk_mutations/products.py:434]: This the case that I explained to you on the call.  You're calling `lambda` with an instance which you get by iterating on something. `On commit` will be called only with the last `instance` object.

File: saleor/graphql/product/mutations/products.py (hunk 14 above)
Comment (korycins) [reply saleor/graphql/product/mutations/products.py:1507]: You're creating transaction only to call  on_commit inside. See my previous comment.

File: saleor/graphql/product/mutations/products.py [hunk 17]
```diff
@@ -1530,4 +1555,10 @@ def perform_mutation(cls, _root, info, image_id, variant_id):
             variant_image.delete()
 
         variant = ChannelContext(node=variant, channel_slug=None)
+
+        with transaction.atomic():
+            transaction.on_commit(
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:1560]: Incorrect use of transaction. See previous comments.

File: saleor/webhook/tests/test_webhook_payloads.py [hunk 18]
```diff
@@ -82,3 +90,62 @@ def test_order_lines_have_all_required_fields(order, order_line_with_one_allocat
...
+        "price_amount": "10.000",
+        "type": "ProductVariantChannelListing",
+    }
+    assert len(payload.keys()) == len(payload_fields)
+
+
+def test_generate_product_variant_deleted_payload(
+    product_with_variant_with_two_attributes,
+):
+    variant = product_with_variant_with_two_attributes.variants.prefetch_related(
```
Comment (korycins) [inline saleor/webhook/tests/test_webhook_payloads.py:129]: IMO to proper simulate the case we should call .delete() in the same way as we did in mutation responsible for  `.delete` action.
@fowczarek  What do you think? 
Here we already fetched the object, so even if we call delete on QS, we will have all details stored locally. It is little bit different way of deleting than we currently use

Comment (piotrgrundas) [reply saleor/webhook/tests/test_webhook_payloads.py:129]: In mutation it's handled the sam way. Object/s is/r being cached before actually being removed, and then
 passed to the event.

File: saleor/graphql/translations/mutations.py [hunk 19]
```diff
@@ -149,6 +151,11 @@ def perform_mutation(cls, _root, info, **data):
             language_code=data["language_code"], defaults=data["input"]
         )
         variant = ChannelContext(node=variant, channel_slug=None)
+
+        transaction.on_commit(
+            lambda: info.context.plugins.product_variant_deleted(variant)
```
Comment (korycins) [inline saleor/graphql/translations/mutations.py:156]: Why delete?

**Review by korycins — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

**Review by fowczarek — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Pull request merged**